compiler/generator_rust.c

```diff
@@ -0,0 +1,1347 @@
+        case c_mathassign:    generate_integer_assign(g, p, "="); break;
+        case c_plusassign:    generate_integer_assign(g, p, "+="); break;
+        case c_minusassign:   generate_integer_assign(g, p, "-="); break;
+        case c_multiplyassign:generate_integer_assign(g, p, "*="); break;
+        case c_divideassign:  generate_integer_assign(g, p, "/="); break;
+        case c_eq:            generate_integer_test(g, p, "=="); break;
+        case c_ne:            generate_integer_test(g, p, "!="); break;
+        case c_gr:            generate_integer_test(g, p, ">"); break;
+        case c_ge:            generate_integer_test(g, p, ">="); break;
+        case c_ls:            generate_integer_test(g, p, "<"); break;
+        case c_le:            generate_integer_test(g, p, "<="); break;
+        case c_call:          generate_call(g, p); break;
+        case c_grouping:      generate_grouping(g, p, false); break;
+        case c_non:           generate_grouping(g, p, true); break;
+        case c_name:          generate_namedstring(g, p); break;
+        case c_literalstring: generate_literalstring(g, p); break;
+        case c_among:         generate_among(g, p); break;
+        case c_substring:     generate_substring(g, p); break;
+        case c_booltest:      generate_booltest(g, p); break;
+        case c_false:         generate_false(g, p); break;
+        case c_true:          break;
+        case c_debug:         generate_debug(g, p); break;
+        default: fprintf(stderr, "%d encountered\n", p->type);
+                 exit(1);
+    }
+
+    g->failure_label = a0;
+    str_delete(g->failure_str);
+    g->failure_str = a1;
+}
+
+static void generate_start_comment(struct generator * g) {
+
+    w(g, "//! This file was generated automatically by the Snowball to Rust compiler~N");
+    w(g, "//! http://snowballstem.org/~N~N");
+}
+
+static void generate_class_begin(struct generator * g) {
+  w(g, "use snowball::SnowballEnv;~N");
+  w(g, "use snowball::Among;~N~N");  
+}
+
+static void generate_among_table(struct generator * g, struct among * x) {
+
+    struct amongvec * v = x->b;
+
+    g->I[0] = x->number;
+    g->I[1] = x->literalstring_count;
+
+    w(g, "~Mstatic A_~I0: &'static [Among; ~I1] = &[~N~+");
+    {
+        int i;
+        for (i = 0; i < x->literalstring_count; i++)
+        {
+            g->I[0] = i;
+            g->I[1] = v->i;
+            g->I[2] = v->result;
+            g->L[0] = v->b;
+            g->S[0] = i < x->literalstring_count - 1 ? "," : "";
+
+            w(g, "~MAmong(~L0, ~I1, ~I2, None");
+            if (v->function != 0)
+            {
+                w(g, ", \"");
+                if (v->function->type == t_routine) {
+                    /* Need to use mangled version of private name here. */
+                    w(g, "_~n");
+                }
+                write_varname(g, v->function);
+                w(g, "\"");
+            }
```

Comment:
Representing functions with mangled strings is unnecessary in Rust. It would be more idiomatic to replace the contents of this block with:
```c
                w(g, ", ");
                write_varname(g, v->function);
```